R/paginate.R

```diff
@@ -0,0 +1,331 @@
+                         total = NULL) {
+  check_request(req)
+  check_function(next_request)
+  check_param(page_size, allow_null = TRUE)
+  check_character(total)
+
+  req <- req_set_param(req, page_size)
+
+  req_policies(
+    req,
+    next_request = next_request,
+    page_size = page_size$value,
+    total = total
+  )
+}
+
+#' Perform a paginated request
+#'
+#' @inheritParams req_perform
+#' @param resp An HTTP [response].
+#' @param max_pages The maximum number of pages to request.
+#' @param progress Display a progress bar?
+#'
+#' @return A list of responses.
+#' @export
+#'
+#' @examples
+#' req_pokemon <- request("https://pokeapi.co/api/v2/pokemon") %>%
+#'   req_paginate_next_url(
+#'     "next",
+#'     page_size = in_query("limit", 150L),
+#'     total = "count"
+#'   )
+#'
+#' responses <- paginate_perform(req_pokemon)
+paginate_perform <- function(req,
+                             max_pages = 20L,
+                             progress = TRUE) {
+  check_request(req)
+  check_bool(progress)
+
+  resp <- req_perform(req)
+
+  n_pages <- paginate_n_pages(resp, req, max_pages = max_pages)
+
+  out <- vector("list", length = n_pages)
+  out[[1]] <- resp
+
+  cli::cli_progress_bar(
+    "Paginate",
+    total = n_pages,
+    format = "{cli::pb_spin} Page {cli::pb_current}/{cli::pb_total} | ETA: {cli::pb_eta}",
+    current = 1L
+  )
+
+  for (page in seq2(2, n_pages)) {
+    req <- paginate_next_request(resp, req)
+    if (is.null(req)) {
+      page <- page - 1L
+      break
+    }
+
+    resp <- req_perform(req)
+
+    body_parsed <- resp_body_json(resp)
+    out[[page]] <- resp
+
+    cli::cli_progress_update()
+  }
+  cli::cli_progress_done()
+
+  # remove unused end of `out` in case the pagination loop exits before all
+  # `max_pages` is reached
+  if (page < n_pages) {
+    out <- out[seq2(1, page)]
+  }
+
+  out
+}
+
+#' @rdname paginate_perform
+paginate_next_request <- function(resp, req) {
+  check_response(resp)
+  check_request(req)
+
+  req_policy_call(
+    req,
+    "next_request",
+    args = list(resp = resp, req = req),
+    default = NULL
+  )
+}
+
+#' @rdname paginate_perform
+paginate_n_pages <- function(resp, req, max_pages) {
+  check_response(resp)
+  check_request(req)
+
+  page_size <- req$policies$page_size
+  total <- req$policies$total
+
+  if (is.null(total) || is.null(page_size)) {
+    return(max_pages)
+  }
+
+  body_parsed <- resp_body_json(resp)
+  total <- purrr::pluck(body_parsed, total)
+  if (is.null(total)) {
+    return(max_pages)
+  }
+  n_pages <- ceiling(total / page_size)
+
+  min(n_pages, max_pages)
+}
+
+#' @rdname req_paginate
+#' @export
+req_paginate_next_url <- function(req,
+                                  next_url,
+                                  ...,
+                                  page_size = NULL,
+                                  total = NULL) {
+  check_character(next_url)
+
+  next_request <- function(req, resp) {
+    body_parsed <- resp_body_json(resp)
```

Comment:
It would be nice to cache the parsed body. Otherwise the body is usually parsed twice.